charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,264 @@
-                        "imageDetails": image_details,
-                        "envConfig": {
-                            "USERID_HEADER": "kubeflow-userid",
-                            "USERID_PREFIX": "",
-                            "APP_PREFIX": "/tensorboards",
-                            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
-                            "BACKEND_MODE": config["backend-mode"],
-                        },
-                        "ports": [{"name": "http", "containerPort": config["port"]}],
-                    }
-                ],
             },
-        )
+        }
+
+        return Layer(layer_config)
+
+    def _on_install(self, _) -> None:
+        """Perform installation only actions."""
+        try:
+            # Deploy K8S resources to speed up deployment
+            self._deploy_k8s_resources()
+        except ErrorWithStatus as err:
+            self._log_and_set_status(err.status)
+        return
+
+    def _on_remove(self, _) -> None:
+        """Remove all resources."""
+        self._log_and_set_status(MaintenanceStatus("Removing K8S resources"))
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as err:
+            # Do not log/report when resources were not found
+            if err.status.code != 404:
+                self.logger.error(f"Removing K8S resources failed with error: {err}")
+                raise GenericCharmRuntimeError("Removing K8s resources failed") from err
+
+    def _on_event(self, event) -> None:
+        """Perform required actions for every event."""
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+            update_layer(
+                self._container_name,
+                self._container,
+                self._tensorboards_web_app_layer,
+                self.logger,
+            )
+            interfaces = self._get_interfaces()
+            self._configure_mesh(interfaces)
+
+        except ErrorWithStatus as err:
+            self._log_and_set_status(err.status)
+            return
 
-        self.model.unit.status = ActiveStatus()
+        self.unit.status = ActiveStatus()
 
-    def _configure_mesh(self, interfaces):
+    def _deploy_k8s_resources(self) -> None:
+        """Deploy K8S resources."""
+        try:
+            self._log_and_set_status(MaintenanceStatus("Creating K8S resources"))
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            raise ErrorWithStatus("Creating K8S resources failed", BlockedStatus)
+        self._log_and_set_status(MaintenanceStatus("K8S resources created"))
+
+    def _is_container_ready(self) -> bool:
+        """Check if connection can be made with container.
+        Set maintenance status if container is not available.
+
+        Return:
```

Comment:
I'd prefer if we raised instead of return, that way you can stop the execution when the container is not ready. Check [this](https://github.com/canonical/kubeflow-tensorboards-operator/blob/main/charms/tensorboard-controller/src/charm.py#L196) example.